### DISCUSSION TITLE
Allowing dependencies besides those implied by function args in the delayed interface

### DISCUSSION THREAD

**Discussion by ghost** [dask/dask#7901] [Ideas]
As far as I'm aware you can only apply dependencies that are inherently implied through a functions arguments. It is potentially contrary to the computational nature of the dask library, but there could be (and are depending on how you use it) cases where you would want to depend on a task for its side-effects, rather than return value.

You can see this suggestion in action in the `aiodag` library. This library provides an interface heavily influenced by dask.delayed to implicitly build dags powered by asyncio.

https://github.com/aa1371/aiodag#endogenous-vs-exogenous-depedencies

An example of the interface extrapolated to dask.delayed is as follows:
```
def task1():
    pass

def task2():
    pass

def task3(x):
    pass

t1 = delayed(task1)()
t2 = delayed(task2, t1)()
t3 = delayed(task3, t1)(t2)
```
In this case task1 is applied as a dependency to task2 and task3, even though neither expects or uses its result.

I briefly tried to search for  any related issues open or closed, but was quite a bit to digest and couldn't find anything. Would there be any appetite for this kind of (or some similar) interface?

**Comment by mrocklin** [comment]
When I need to do this I tend to pass along some token between the different tasks

```python
@dask.delayed
def task1():
    pass

@dask.delayed
def task2(extra=None):
    pass

@dask.delayed
def task3(extra=None):
    pass

@dask.delayed
def token(x):
    return None

t1 = task1()
t2 = task2(extra=token(t1))
t3 = task3(t2, extra=token(t1))
```

We don't pass any real data (just None) and it's easy enough to reuse the existing system.

It's not exactly what you want, but it might do

**Reply by abergou** [reply]
I think the dask.graph_manipulation sub module also contains utilities for this.

**Reply by ghost** [reply]
@abergou thanks for the suggestion, looks like `dask.graph_manipulation.bind` does the trick perfectly.
```
from dask import delayed
from dask.graph_manipulation import bind

def test1():
    print('test1')

def test2():
    print('test2')

def test3(val):
    print('test3')

t1 = delayed(test1)()
t2 = bind(delayed(test2), t1)()
t3 = bind(delayed(test3), t1)(t2)
```

@mrocklin Think it might still be worth doing something like adding a `bind` parameter to `delayed` which will then apply the bind, just to simplify the invocation a bit.
